### PR TITLE
fail divergent ready nodes

### PR DESCRIPTION
**Pull request by andreer**

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeFailer.java [hunk 1]
```diff
@@ -130,6 +134,13 @@ private void updateNodeLivenessEventsForReadyNodes() {
                 .collect(Collectors.toList());
     }
 
+    private List<Node> readyNodesWithHardwareDivergence() {
+        return nodeRepository().getNodes(Node.State.ready).stream()
+                .filter(node -> node.status().hardwareDivergence().isPresent()
+                        && !node.status().hardwareDivergence().get().equals("null"))
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeFailer.java]: Can the divergence field really contain the string "null"? If so can we please change `NodeSerializer` to throw away values that are "null" and have `NodePatcher.asOptionalString` filter incoming "null" values?

I assume it happens if clients send `{"hardwareDivergence": "null"}` instead of `{"hardwareDivergence": null}` as the latter is handled.

**Review by mpolden — CHANGES_REQUESTED** [review]

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeFailer.java (hunk 1 above)
Comment (andreer) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeFailer.java]: Yes, most nodes today have it set to "null". I guess we should fix the client (SpecVerifier) too.

**Comment by andreer** [comment]
@mpolden The actual reporting seems to be done in chef, so this is the best we can do in this repo. I'll open another PR to fix the chef recipe.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/NodeSerializer.java [hunk 2]
```diff
@@ -239,6 +239,12 @@ private ClusterMembership clusterMembershipFromSlime(Inspector object) {
         return Optional.empty();
     }
 
+    // Remove when we no longer have "null" strings for this field in the node repo
+    private Optional<String> removeQuotedNulls(Optional<String> value) {
+        return value.isPresent() && value.get().equals("null") ? Optional.empty() : value;
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/persistence/NodeSerializer.java]: Can be simplified to `value.filter(v -> !v.equals("null"))`

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodePatcher.java [hunk 3]
```diff
@@ -170,6 +170,11 @@ private String asString(Inspector field) {
         return field.type().equals(Type.NIX) ? Optional.empty() : Optional.of(asString(field));
     }
 
+    // Remove when we no longer have "null" strings for this field in the node repo
+    private Optional<String> removeQuotedNulls(Optional<String> value) {
+        return value.isPresent() && value.get().equals("null") ? Optional.empty() : value;
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodePatcher.java]: Same as above.

**Review by mpolden — APPROVED** [review]
LGTM. One minor comment.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodePatcher.java (hunk 3 above)
Comment (mpolden) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodePatcher.java]: Also, it would be nice have test that sends "null" at the REST API level.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeFailer.java (hunk 1 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeFailer.java]: The problem is that `node-admin`'s node-repo client doesn't support updating value to `null` in its current state (because of `JsonInclude.Include.NON_NULL`), so we had to improvise a little as the students were on their final week. I dont think anyone has touched that code since :man_shrugging: .

**Comment by freva** [comment]
@andreer Yes, for docker nodes it's updated here: https://github.com/vespa-engine/vespa/blob/aea153996afeb9c1b5c3772adaa188d47b4284da/node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/nodeadmin/NodeAdminStateUpdater.java#L126

**Comment by mpolden** [comment]
LGTM

**Comment by freva** [comment]
Can we also add a condition in `FailedExpirer` to moved nodes with spec divergence to `parked` so it's not stuck in a `failed`-`dirty` loop?

**Comment by mpolden** [comment]
@freva Agree.

All the exceptions in `FailedExpirer` should be moved to a separate method, e.g. `canRecycle` or similar. Feel free to do that in a separate PR though.

**Comment by andreer** [comment]
Thanks for catching the failed/dirty loop. I piggybacked it on parking of hardwarefailured nodes, as I think we'd want to have the same logic if it happens to a docker host.

**Comment by mpolden** [comment]
I still think the `FailedExpirer.expire` method could use some refactoring, but this is fine for now. 👍
